// @angular-extensions/elements/esm2015/lib/lazy-elements/lazy-element/lazy-element.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
import { LazyElementsLoaderService } from '../lazy-elements-loader.service';
/** @type {?} */
const LOG_PREFIX = '@angular-extensions/elements';
export class LazyElementDirective {
    // tslint:disable-line:no-input-rename
    /**
     * @param {?} vcr
     * @param {?} template
     * @param {?} elementsLoaderService
     */
    constructor(vcr, template, elementsLoaderService) {
        this.vcr = vcr;
        this.template = template;
        this.elementsLoaderService = elementsLoaderService;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        /** @type {?} */
        const elementTag = ((/** @type {?} */ (this.template)))._def.element.template.nodes[0]
            .element.name;
        if (this.loadingTemplateRef) {
            this.vcr.createEmbeddedView(this.loadingTemplateRef);
        }
        this.elementsLoaderService
            .loadElement(this.url, elementTag, this.isModule)
            .then((/**
         * @return {?}
         */
        () => {
            this.vcr.clear();
            this.vcr.createEmbeddedView(this.template);
        }))
            .catch((/**
         * @return {?}
         */
        () => {
            this.vcr.clear();
            if (this.errorTemplateRef) {
                this.vcr.createEmbeddedView(this.errorTemplateRef);
            }
            else {
                console.error(`${LOG_PREFIX} - Loading of element <${elementTag}> failed, please provide <ng-template #error>Loading failed...</ng-template> and reference it in *axLazyElement="errorTemplate: error" to display customized error message in place of element`);
            }
        }));
    }
}
LazyElementDirective.decorators = [
    { type: Directive, args: [{
                selector: '[axLazyElement]'
            },] }
];
/** @nocollapse */
LazyElementDirective.ctorParameters = () => [
    { type: ViewContainerRef },
    { type: TemplateRef },
    { type: LazyElementsLoaderService }
];
LazyElementDirective.propDecorators = {
    url: [{ type: Input, args: ['axLazyElement',] }],
    loadingTemplateRef: [{ type: Input, args: ['axLazyElementLoadingTemplate',] }],
    errorTemplateRef: [{ type: Input, args: ['axLazyElementErrorTemplate',] }],
    isModule: [{ type: Input, args: ['axLazyElementModule',] }]
};
if (false) {
    /** @type {?} */
    LazyElementDirective.prototype.url;
    /** @type {?} */
    LazyElementDirective.prototype.loadingTemplateRef;
    /** @type {?} */
    LazyElementDirective.prototype.errorTemplateRef;
    /** @type {?} */
    LazyElementDirective.prototype.isModule;
    /**
     * @type {?}
     * @private
     */
    LazyElementDirective.prototype.vcr;
    /**
     * @type {?}
     * @private
     */
    LazyElementDirective.prototype.template;
    /**
     * @type {?}
     * @private
     */
    LazyElementDirective.prototype.elementsLoaderService;
}
