// @angular-extensions/elements/esm5/lib/lazy-elements/lazy-element/lazy-element.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Directive, Input, TemplateRef, ViewContainerRef } from '@angular/core';
import { LazyElementsLoaderService } from '../lazy-elements-loader.service';
/** @type {?} */
var LOG_PREFIX = '@angular-extensions/elements';
var LazyElementDirective = /** @class */ (function () {
    function LazyElementDirective(vcr, template, elementsLoaderService) {
        this.vcr = vcr;
        this.template = template;
        this.elementsLoaderService = elementsLoaderService;
    }
    /**
     * @return {?}
     */
    LazyElementDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        var _this = this;
        /** @type {?} */
        var elementTag = ((/** @type {?} */ (this.template)))._def.element.template.nodes[0]
            .element.name;
        if (this.loadingTemplateRef) {
            this.vcr.createEmbeddedView(this.loadingTemplateRef);
        }
        this.elementsLoaderService
            .loadElement(this.url, elementTag, this.isModule)
            .then((/**
         * @return {?}
         */
        function () {
            _this.vcr.clear();
            _this.vcr.createEmbeddedView(_this.template);
        }))
            .catch((/**
         * @return {?}
         */
        function () {
            _this.vcr.clear();
            if (_this.errorTemplateRef) {
                _this.vcr.createEmbeddedView(_this.errorTemplateRef);
            }
            else {
                console.error(LOG_PREFIX + " - Loading of element <" + elementTag + "> failed, please provide <ng-template #error>Loading failed...</ng-template> and reference it in *axLazyElement=\"errorTemplate: error\" to display customized error message in place of element");
            }
        }));
    };
    LazyElementDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[axLazyElement]'
                },] }
    ];
    /** @nocollapse */
    LazyElementDirective.ctorParameters = function () { return [
        { type: ViewContainerRef },
        { type: TemplateRef },
        { type: LazyElementsLoaderService }
    ]; };
    LazyElementDirective.propDecorators = {
        url: [{ type: Input, args: ['axLazyElement',] }],
        loadingTemplateRef: [{ type: Input, args: ['axLazyElementLoadingTemplate',] }],
        errorTemplateRef: [{ type: Input, args: ['axLazyElementErrorTemplate',] }],
        isModule: [{ type: Input, args: ['axLazyElementModule',] }]
    };
    return LazyElementDirective;
}());
export { LazyElementDirective };
if (false) {
    /** @type {?} */
    LazyElementDirective.prototype.url;
    /** @type {?} */
    LazyElementDirective.prototype.loadingTemplateRef;
    /** @type {?} */
    LazyElementDirective.prototype.errorTemplateRef;
    /** @type {?} */
    LazyElementDirective.prototype.isModule;
    /**
     * @type {?}
     * @private
     */
    LazyElementDirective.prototype.vcr;
    /**
     * @type {?}
     * @private
     */
    LazyElementDirective.prototype.template;
    /**
     * @type {?}
     * @private
     */
    LazyElementDirective.prototype.elementsLoaderService;
}
